Add service for "plugin" resources

justinas opened the pull request:
Second part of changes initially submitted in https://github.com/gravitational/teleport/pull/20609.

Contributes to https://github.com/gravitational/cloud/issues/2506, see also [Cloud RFD 48](https://github.com/gravitational/cloud/blob/master/rfd/0048-hosted-slack-plugin.md).

Adds service for the previously implemented `Plugin` resource. ~Also adds gRPC request/response types, as at least `CreatePluginRequest` is used here. I did not yet add the gRPC methods themselves as that would automatically require implementations in AuthServer. I plan to submit those in the next PR.~

Commit:
Add Plugins service

Commit:
Add tests for local plugins service

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -0,0 +1,198 @@
+		return trace.BadParameter("req.Plugin is nil")
+	}
+
+	// TODO: exchange initial credentials for live credentials
```
```suggestion
	// TODO(justinas): exchange initial credentials for live credentials
```

What's the plan here? Any sort of active interaction with a network service (like trading some initial credential for a renewable one) should not be done here but rather in `auth.Server`, and here we should only really deal with data (perhaps by taking a `types.Plugin` rather than a `CreatePluginRequest`)

justinas replied:
I agree that auth being responsible for the exchange makes more sense.

I initially extended `proto.CreatePluginRequest` to the service to keep the signature the same as auth, so we can add `services.Plugins` to [ClientI](https://github.com/gravitational/teleport/blob/06bf633ab5cbeca37690b1599fa028f36c474173/lib/auth/clt.go#L1574) and the auth server implements those methods (since signatures match).

We could either:
* keep `CreatePluginRequest` in the service, but assert that the credentials have been exchanged by the caller (this keeps the methods identical and allows interface composition)
* make the service have a `CreatePlugin(types.Plugin)` (which looks cleaner), and manually define a `CreatePlugin(*proto.CreatePluginRequest)` in `ClientI`.

strideynet replied:
I'd prefer option 2 here.

I've so far found that our attempts to keep signatures the same across layers tends to lead to more pain (and only slightly reducing duplication/boilerplate), and I think that this accepting `types.Plugin` avoids us coupling this method to much to the transport layer. This'll make life easier if we introduce a V2 format of the message in future.

This also reduces the likelihood of a refactoring snafu leading to this implementation becoming exposed without sufficient authentication of initial credentials.

espadolini commented on lib/services/local/plugins.go:
```diff
@@ -0,0 +1,198 @@
+	})
+}
+
+func (s *PluginsService) update(ctx context.Context, name string, modify func(types.Plugin) error) error {
```
```suggestion
func (s *PluginsService) updateAndSwap(ctx context.Context, name string, modify func(types.Plugin) error) error {
```

strideynet replied:
Agreed

Commit:
Remove validation calls from service

These are already invoked by MarshalPlugin

Commit:
Separate concerns of local Plugins service

Credentials will be exchanged by the auth server,
so there is no need to have proto.CreatePluginsRequest
in the service layer.

Proto types will be reintroduced in a later PR
covering the auth server changes.

Commit:
Merge remote-tracking branch 'origin/master' into justinas/add-plugin-backend-service

Commit:
Rename method to better reflect purpose

Commit:
Run GCI
